core/drivers/crypto/hisilicon/hpre_main.c

```diff
@@ -0,0 +1,326 @@
+	val |= HPRE_CORE_GATE_ENABLE;
+	io_write32(qm->io_base + HPRE_CORE_SHB_CFG, val);
+}
+
+static int32_t hpre_engine_init(struct acc_device *hpre_dev)
+{
+	struct hisi_qm *qm = &hpre_dev->qm;
+	uint32_t val = 0;
+	int32_t ret;
+
+	if (qm->fun_type == HISI_QM_HW_VF)
+		return 0;
+
+	hpre_disable_clock_gate(qm);
+	hisi_qm_dev_init(qm);
+
+	io_write32(qm->io_base + HPRE_CFG_AXCACHE, HPRE_AXCACHE_MASK);
+	io_write32(qm->io_base + HPRE_BD_ENDIAN, HPRE_BD_LITTLE_ENDIAN);
+	io_write32(qm->io_base + HPRE_RAS_CE_ENB, HPRE_RAS_CE_MASK);
+	io_write32(qm->io_base + HPRE_RAS_NFE_ENB, HPRE_RAS_NFE_MASK);
+	io_write32(qm->io_base + HPRE_RAS_FE_ENB, HPRE_RAS_FE_MASK);
+	io_write32(qm->io_base + HPRE_HAC_INT_MASK, HPRE_HAC_INT_DISABLE);
+	io_write32(qm->io_base + HPRE_POISON_BYPASS, 0x0);
+	io_write32(qm->io_base + HPRE_COMMON_CNT_CLR_CE, 0x0);
+	io_write32(qm->io_base + HPRE_ECC_BYPASS, 0x0);
+	/* cmd_type is controlled by hac subctrl */
+	io_write32(qm->io_base + HPRE_BD_ARUSR_CFG, HPRE_BD_ARUSR_MASK);
+	io_write32(qm->io_base + HPRE_BD_AWUSR_CFG, HPRE_BD_AWUSR_MASK);
+	io_write32(qm->io_base + HPRE_DATA_RUSER_CFG, HPRE_DATA_USR_MASK);
+	io_write32(qm->io_base + HPRE_DATA_WUSER_CFG, HPRE_DATA_USR_MASK);
+	io_write32(qm->io_base + HPRE_TYPES_ENB, HPRE_RSA_ENB | HPRE_ECC_ENB);
+	io_write32(qm->io_base + HPRE_RDCHN_INI_CFG, 0x1);
+	ret = IO_READ32_POLL_TIMEOUT(qm->io_base + HPRE_RDCHN_INI_ST, val,
+				     val & 0x1, POLL_PERIOD, POLL_TIMEOUT);
+	if (ret) {
+		EMSG("Fail to init rd channel");
+		return ret;
+	}
+
+	ret = hpre_set_cluster(qm);
+	if (ret) {
+		EMSG("Fail to init hpre cluster cores");
+		return ret;
+	}
+
+	hpre_enable_clock_gate(qm);
+
+	return 0;
+}
+
+static enum hisi_drv_status hpre_dev_status_check(struct hisi_qm *qm)
+{
+	uint32_t val = 0;
+
+	val = io_read32(qm->io_base + HPRE_HAC_INT_SRC);
+	if (val & HPRE_RAS_NFE_MASK) {
+		EMSG("HPRE NFE RAS happened, need to reset");
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	val = io_read32(qm->io_base + HISI_QM_ABNML_INT_SRC);
+	if (val) {
+		if (val & HISI_QM_HPRE_NFE_INT_MASK)
+			EMSG("QM NFE RAS happened, need to reset");
+
+		if (val & HISI_QM_INVALID_DB) {
+			EMSG("QM invalid db happened, please check");
+			io_write32(qm->io_base + HISI_QM_ABNML_INT_SRC,
+				   HISI_QM_INVALID_DB);
+		}
+
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	return 0;
+}
+
+static enum hisi_drv_status hpre_qm_init(struct acc_device *hpre_dev)
+{
+	struct hisi_qm *qm = &hpre_dev->qm;
+
+	if (cpu_mmu_enabled()) {
+		qm->io_base = (uintptr_t)phys_to_virt_io(hpre_dev->io_base,
+							 hpre_dev->io_size);
+		if (!qm->io_base) {
+			EMSG("Fail to get qm io_base");
+			return HISI_QM_DRVCRYPT_EFAULT;
+		}
+	} else {
+		qm->io_base = hpre_dev->io_base;
+	}
+
+	qm->vfs_num = hpre_dev->vfs_num;
+	qm->fun_type = hpre_dev->fun_type;
+	qm->sqe_size = HPRE_SQE_SIZE;
+	qm->sqe_log2_size = HPRE_SQE_LOG2_SIZE;
+	if (qm->fun_type == HISI_QM_HW_PF) {
+		hisi_qm_get_version(qm);
+		IMSG("HPRE hardware version is 0x%x", qm->version);
```

Comment:
`qm->version` is of type `uint32_t` hence: `"HPRE hardware version is 0x%"PRIx32, qm->version)`